### PR TITLE
fix(positioning): use CSS translate for positioning

### LINKED ISSUE #2470: Tooltip placement incorrect when element is near edge and when using ng-template
### Bug description:

So I had reported #2465 earlier, but it turns out my issue was something else entirely. If you are using an ng-template to display the content of a tooltip, and the element that displays the tooltip is near the edge of the screen, the tooltip will be incorrectly placed. You can see in the below stackblitz that there are three buttons. All three pull from the same variable for it's content. But the last two, display the content in a ng-template. The second button works fine, but the third button which is at the edge of the screen you can see is positioned incorrectly, there is plenty of room for it all around, but its offset from the element incorrectly.

### Link to minimally-working plunker that reproduces the issue:

https://stackblitz.com/edit/angular-fneazi-ljhkng?file=app%2Ftooltip-basic.html

### Version of Angular, ng-bootstrap, and Bootstrap:

Angular: 6.0.0

ng-bootstrap: 2.1.1

Bootstrap: 4.1.0

### PR DESCRIPTION
**Pull request by fbasso**
fix #2470

### CODE REVIEW

**Comment by maxokorokov** [comment]
It breaks this case for me:

screen shot 2019-01-16 at 14 14 37

**Comment by fbasso** [comment]
ok, it seems more complicated than expected. I'll try a complete refactor of the positioning utility.

**Comment by fbasso** [comment]
This is a new version . It includes a refactor about how the positioning is calculated. 

The positioning was pre-calculated before applying the positions. In some edge case, the width or height of the element could change afterwards, leading to issues.

Now the positioning is really done for each wanted position (i.e. 'top', 'right', ...) until the element is inside the viewport or the last asked position.

File: src/tooltip/tooltip.spec.ts [hunk 1]
```diff
@@ -28,15 +28,14 @@ describe('ngb-tooltip-window', () => {
     fixture.detectChanges();
 
     expect(fixture.nativeElement).toHaveCssClass('tooltip');
-    expect(fixture.nativeElement).toHaveCssClass('bs-tooltip-top');
+    expect(fixture.nativeElement).not.toHaveCssClass('bs-tooltip-top');
     expect(fixture.nativeElement.getAttribute('role')).toBe('tooltip');
   });
 
-  it('should position tooltips as requested', () => {
+  it(`shouldn't add 'bs-tooltip-left' class with binding`, () => {
```
Comment (maxokorokov) [inline src/tooltip/tooltip.spec.ts]: Why should it even try applying left if you removed `fixture.componentInstance.placement = 'left';` ?

this isn't testing anything now, you should put it back :)

See what you did in `popover.spec.ts`

File: src/popover/popover.ts [hunk 2]
```diff
@@ -102,12 +84,12 @@ export class NgbPopover implements OnInit, OnDestroy, OnChanges {
...
-   *    "left", "left-top", "left-bottom", "right", "right-top", "right-bottom"
-   * and array of above values.
-   */
-  @Input() placement: PlacementArray;
+    * Placement of a tooltip accepts:
+    *  - Array or space separated string of following values:
+    *    "top", "top-left", "top-right", "bottom", "bottom-left", "bottom-right",
+    *    "left", "left-top", "left-bottom", "right", "right-top", "right-bottom",
+    */
+  @Input() placement: string | PlacementArray;
```
Comment (maxokorokov) [inline src/popover/popover.ts]: Let's open a separate feature PR for this later for ALL widgets supporting positioning

File: src/tooltip/tooltip.ts [hunk 3]
```diff
@@ -83,11 +62,11 @@ export class NgbTooltip implements OnInit, OnDestroy {
...
   /**
     * Placement of a tooltip accepts:
+    *  - Array or space separated string of following values:
     *    "top", "top-left", "top-right", "bottom", "bottom-left", "bottom-right",
-    *    "left", "left-top", "left-bottom", "right", "right-top", "right-bottom"
-    * and array of above values.
+    *    "left", "left-top", "left-bottom", "right", "right-top", "right-bottom",
     */
-  @Input() placement: PlacementArray;
+  @Input() placement: string | PlacementArray;
```
Comment (maxokorokov) [inline src/tooltip/tooltip.ts]: Let's open a separate feature PR for this later for ALL widgets supporting positioning

File: src/tooltip/tooltip.ts [hunk 4]
```diff
@@ -178,18 +156,14 @@ export class NgbTooltip implements OnInit, OnDestroy {
         this._document.querySelector(this.container).appendChild(this._windowRef.location.nativeElement);
       }
 
-      this._windowRef.instance.placement = Array.isArray(this.placement) ? this.placement[0] : this.placement;
+      if (!Array.isArray(this.placement)) {
```
Comment (maxokorokov) [inline src/tooltip/tooltip.ts]: Let's open a separate feature PR for this later for ALL widgets supporting positioning

File: src/tooltip/tooltip.spec.ts [hunk 5]
```diff
@@ -292,6 +291,34 @@ describe('ngb-tooltip', () => {
         expect(windowEl.textContent.trim()).toBe('Great tip!');
       });
 
+      it('should accept placement in space seperated (second value should be applied)', () => {
```
Comment (maxokorokov) [inline src/tooltip/tooltip.spec.ts]: Let's open a separate feature PR for this later for ALL widgets supporting positioning

File: src/popover/popover.ts [hunk 6]
```diff
@@ -61,19 +56,6 @@ export class NgbPopoverWindow {
   constructor(private _element: ElementRef<HTMLElement>, private _renderer: Renderer2) {}
```
Comment (maxokorokov) [inline src/popover/popover.ts]: you should remove the constructor as well, not used anymore

File: e2e-app/src/app/tooltip/position/tooltip-position.component.html [hunk 7]
```diff
@@ -0,0 +1,31 @@
+<h3>Tooltip autoclose tests</h3>
```
Comment (maxokorokov) [inline e2e-app/src/app/tooltip/position/tooltip-position.component.html]: Tooltip positioning tests

File: e2e-app/src/app/tooltip/position/tooltip-position.component.html [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+<form id="default">
+  <div class="m-3">
+    <button class="btn btn-outline-secondary mx-1" id="flex-left" (click)="setPosition('start')">Left</button>
+    <button class="btn btn-outline-secondary mx-1" id="flex-center" (click)="setPosition('center')">Center</button>
+    <button class="btn btn-outline-secondary mx-1" id="flex-right" (click)="setPosition('end')">Right</button>
+  </div>
+  <div class="d-flex" [ngClass]="flexPosition">
```
Comment (maxokorokov) [inline e2e-app/src/app/tooltip/position/tooltip-position.component.html]: nitpick: could simply be 
```html
<div class="d-flex {{flexPosition}}">
```

File: e2e-app/src/app/tooltip/position/tooltip-position.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
+import {Key, browser} from 'protractor';
```
Comment (maxokorokov) [inline e2e-app/src/app/tooltip/position/tooltip-position.e2e-spec.ts]: please cleanup unused imports

File: e2e-app/src/app/tooltip/position/tooltip-position.po.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import {$} from 'protractor';
+
+export class TooltipPositionPage {
+  async clickOutside() { await $('#outside-button').click(); }
```
Comment (maxokorokov) [inline e2e-app/src/app/tooltip/position/tooltip-position.po.ts]: please cleanup unused methods

**Comment by fbasso** [comment]
I think the code is almost ready now. I'll try to find out what are the issues this PR will fix.

**Comment by fbasso** [comment]
fix #2914, but placement alternatives need to be specified (otherwise, the tooltip is not completely inside the viewport). i.e. : `[placement]="['top', 'auto']"`

**Comment by fbasso** [comment]
close #2069 
The same, placements alternatives need to be specified to get it into the viewport. 

The main idea is if only one placement is specified, it will stick on it even if it's partly outside of the view, but it will never overlap the main element (from which the tooltip/popover is triggered).

**Comment by fbasso** [comment]
close #2924

**Comment by fbasso** [comment]
Partially fix #2346. The popover/dropdown content is no longer resized, but is still arrow positioning issue. For example, if the placement ends to be on bottom-right, the arrow will be 2rem from the right, like this:

image

I'm not sure ng-bootstrap will be able to manage this case, styling from the application will be necessary (for example, override .bs-popover-bottom-right .arrow)

**Pull request merged**